s390x build fails due various MONO_OFFSET_MonoLMF_* undeclared

Since commit 79135c6ea57b516addf187a08cb01826bb0197c6 the s390x build fails with
````
make[4]: Entering directory '/home/jenkins/workspace/Mono/label/s390x/mono/mini'
  CC       libmini_la-mini-s390x.lo
mini-s390x.c: In function ‘mono_arch_emit_call’:
mini-s390x.c:1913:12: warning: variable ‘size’ set but not used [-Wunused-but-set-variable]
 1913 |    guint32 size;
      |            ^~~~
In file included from ../../mono/metadata/object-internals.h:19,
                 from mini.h:28,
                 from mini-s390x.c:339:
mini-s390x.c: In function ‘mono_arch_emit_prolog’:
../../mono/metadata/abi-details.h:68:43: error: ‘MONO_OFFSET_MonoLMF_pregs’ undeclared (first use in this function); did you mean ‘MONO_OFFSET_MonoLMF_previous_lmf’?
   68 | #define MONO_STRUCT_OFFSET(struct,field) (MONO_OFFSET_ ## struct ## _ ## field == -1, G_STRUCT_OFFSET (struct,field))
      |                                           ^~~~~~~~~~~~
mini-s390x.c:5642:51: note: in expansion of macro ‘MONO_STRUCT_OFFSET’
 5642 |   mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, pregs[0]), SLOT_NOREF);
      |                                                   ^~~~~~~~~~~~~~~~~~
../../mono/metadata/abi-details.h:68:43: note: each undeclared identifier is reported only once for each function it appears in
   68 | #define MONO_STRUCT_OFFSET(struct,field) (MONO_OFFSET_ ## struct ## _ ## field == -1, G_STRUCT_OFFSET (struct,field))
      |                                           ^~~~~~~~~~~~
mini-s390x.c:5642:51: note: in expansion of macro ‘MONO_STRUCT_OFFSET’
 5642 |   mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, pregs[0]), SLOT_NOREF);
      |                                                   ^~~~~~~~~~~~~~~~~~
../../mono/metadata/abi-details.h:68:43: error: ‘MONO_OFFSET_MonoLMF_lmf_addr’ undeclared (first use in this function); did you mean ‘MONO_OFFSET_MonoFtnDesc_addr’?
   68 | #define MONO_STRUCT_OFFSET(struct,field) (MONO_OFFSET_ ## struct ## _ ## field == -1, G_STRUCT_OFFSET (struct,field))
      |                                           ^~~~~~~~~~~~
mini-s390x.c:5661:51: note: in expansion of macro ‘MONO_STRUCT_OFFSET’
 5661 |   mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, lmf_addr), SLOT_NOREF);
      |                                                   ^~~~~~~~~~~~~~~~~~
../../mono/metadata/abi-details.h:68:43: error: ‘MONO_OFFSET_MonoLMF_method’ undeclared (first use in this function); did you mean ‘MONO_OFFSET_MonoDelegate_method’?
   68 | #define MONO_STRUCT_OFFSET(struct,field) (MONO_OFFSET_ ## struct ## _ ## field == -1, G_STRUCT_OFFSET (struct,field))
      |                                           ^~~~~~~~~~~~
mini-s390x.c:5686:51: note: in expansion of macro ‘MONO_STRUCT_OFFSET’
 5686 |   mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, method), SLOT_NOREF);
      |                                                   ^~~~~~~~~~~~~~~~~~
../../mono/metadata/abi-details.h:68:43: error: ‘MONO_OFFSET_MonoLMF_ebp’ undeclared (first use in this function); did you mean ‘MONO_OFFSET_MonoVTable_type’?
   68 | #define MONO_STRUCT_OFFSET(struct,field) (MONO_OFFSET_ ## struct ## _ ## field == -1, G_STRUCT_OFFSET (struct,field))
      |                                           ^~~~~~~~~~~~
mini-s390x.c:5694:51: note: in expansion of macro ‘MONO_STRUCT_OFFSET’
 5694 |   mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, ebp), SLOT_NOREF);
      |                                                   ^~~~~~~~~~~~~~~~~~
../../mono/metadata/abi-details.h:68:43: error: ‘MONO_OFFSET_MonoLMF_eip’ undeclared (first use in this function); did you mean ‘MONO_OFFSET_MonoVTable_type’?
   68 | #define MONO_STRUCT_OFFSET(struct,field) (MONO_OFFSET_ ## struct ## _ ## field == -1, G_STRUCT_OFFSET (struct,field))
      |                                           ^~~~~~~~~~~~
mini-s390x.c:5695:51: note: in expansion of macro ‘MONO_STRUCT_OFFSET’
 5695 |   mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, eip), SLOT_NOREF);
      |                                                   ^~~~~~~~~~~~~~~~~~
../../mono/metadata/abi-details.h:68:43: error: ‘MONO_OFFSET_MonoLMF_gregs’ undeclared (first use in this function); did you mean ‘MONO_OFFSET_MonoLMF_previous_lmf’?
   68 | #define MONO_STRUCT_OFFSET(struct,field) (MONO_OFFSET_ ## struct ## _ ## field == -1, G_STRUCT_OFFSET (struct,field))
      |                                           ^~~~~~~~~~~~
mini-s390x.c:5702:51: note: in expansion of macro ‘MONO_STRUCT_OFFSET’
 5702 |   mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, gregs[0]), SLOT_NOREF);
      |                                                   ^~~~~~~~~~~~~~~~~~
../../mono/metadata/abi-details.h:68:43: error: ‘MONO_OFFSET_MonoLMF_fregs’ undeclared (first use in this function); did you mean ‘MONO_OFFSET_MonoVTable_flags’?
   68 | #define MONO_STRUCT_OFFSET(struct,field) (MONO_OFFSET_ ## struct ## _ ## field == -1, G_STRUCT_OFFSET (struct,field))
      |                                           ^~~~~~~~~~~~
mini-s390x.c:5714:26: note: in expansion of macro ‘MONO_STRUCT_OFFSET’
 5714 |   fpOffset = lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, fregs[0]);
      |                          ^~~~~~~~~~~~~~~~~~
mini-s390x.c: In function ‘mono_arch_emit_exceptions’:
mini-s390x.c:5894:12: warning: variable ‘throw_ip’ set but not used [-Wunused-but-set-variable]
 5894 |    guint64 throw_ip;
      |            ^~~~~~~~
At top level:
mini-s390x.c:740:1: warning: ‘cvtMonoType’ defined but not used [-Wunused-function]
  740 | cvtMonoType(MonoTypeEnum t)
      | ^~~~~~~~~~~
mini-s390x.c:716:1: warning: ‘indent’ defined but not used [-Wunused-function]
  716 | indent (int diff) {
      | ^~~~~~
mini-s390x.c:655:1: warning: ‘retFitsInReg’ defined but not used [-Wunused-function]
  655 | retFitsInReg(guint32 size)
      | ^~~~~~~~~~~~
make[4]: *** [Makefile:2515: libmini_la-mini-s390x.lo] Error 1

````

I haven't looked at the details yet, maybe there is simple fix.

CC @nealef

So I went with
````
diff --git a/mono/metadata/object-offsets.h b/mono/metadata/object-offsets.h
index 8e235409969..9a36613866d 100644
--- a/mono/metadata/object-offsets.h
+++ b/mono/metadata/object-offsets.h
@@ -241,6 +241,14 @@ DECL_OFFSET(DynCallArgs, fpregs)
 DECL_OFFSET(DynCallArgs, n_stackargs)
 DECL_OFFSET(DynCallArgs, n_fpargs)
 DECL_OFFSET(DynCallArgs, n_fpret)
+#elif defined(TARGET_S390X)
+DECL_OFFSET(MonoLMF, pregs)
+DECL_OFFSET(MonoLMF, lmf_addr)
+DECL_OFFSET(MonoLMF, method)
+DECL_OFFSET(MonoLMF, ebp)
+DECL_OFFSET(MonoLMF, eip)
+DECL_OFFSET(MonoLMF, gregs)
+DECL_OFFSET(MonoLMF, fregs)
 #endif
 
 // Shared architecture offfsets
````
but now I'm getting
````
In file included from ../../mono/metadata/object-internals.h:19,
                 from mini.h:28,
                 from mini-s390x.c:339:
mini-s390x.c: In function ‘mono_arch_emit_prolog’:
mini-s390x.c:5642:85: error: subscripted value is neither array nor pointer nor vector
 5642 |   mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, pregs[0]), SLOT_NOREF);
      |                                                                                     ^
../../mono/metadata/abi-details.h:68:74: note: in definition of macro ‘MONO_STRUCT_OFFSET’
   68 | #define MONO_STRUCT_OFFSET(struct,field) (MONO_OFFSET_ ## struct ## _ ## field == -1, G_STRUCT_OFFSET (struct,field))
      |                                                                          ^~~~~
mini-s390x.c:5643:85: error: subscripted value is neither array nor pointer nor vector
...
````
Looks to me that `MONO_STRUCT_OFFSET` doesn't like accessing array members, while `G_STRUCT_OFFSET` can deal with them. See https://github.com/mono/mono/blob/master/mono/mini/mini-s390x.c#L5627 for the corresponding code.

--- a/src/mono/mono/metadata/object-offsets.h
+++ b/src/mono/mono/metadata/object-offsets.h
@@ -241,6 +241,14 @@ DECL_OFFSET(DynCallArgs, fpregs)
 DECL_OFFSET(DynCallArgs, n_stackargs)
 DECL_OFFSET(DynCallArgs, n_fpargs)
 DECL_OFFSET(DynCallArgs, n_fpret)
+#elif defined(TARGET_S390X)
+DECL_OFFSET(MonoLMF, pregs)
+DECL_OFFSET(MonoLMF, lmf_addr)
+DECL_OFFSET(MonoLMF, method)
+DECL_OFFSET(MonoLMF, ebp)
+DECL_OFFSET(MonoLMF, eip)
+DECL_OFFSET(MonoLMF, gregs)
+DECL_OFFSET(MonoLMF, fregs)
 #endif
 
 // Shared architecture offfsets

--- a/src/mono/mono/mini/mini-s390x.c
+++ b/src/mono/mono/mini/mini-s390x.c
@@ -3722,7 +3722,7 @@ mono_arch_output_basic_block (MonoCompile *cfg, MonoBasicBlock *bb)
 				s390_lgr (code, s390_r14, STK_BASE);
 				s390_sgr (code, s390_r14, s390_r1);
 				s390_stg (code, s390_r14, 0, s390_r13,
-					  G_STRUCT_OFFSET(MonoLMF, ebp));
+					  MONO_STRUCT_OFFSET(MonoLMF, ebp));
                         }
 			s390_lg   (code, s390_r13, 0, STK_BASE, 0);
 			s390_sgr  (code, STK_BASE, s390_r1);
@@ -5637,13 +5637,10 @@ mono_arch_emit_prolog (MonoCompile *cfg)
 		/* Preserve the parameter registers while we fix up the lmf	 */
 		/*---------------------------------------------------------------*/
 		s390_stmg  (code, s390_r2, s390_r6, s390_r13,
-			    G_STRUCT_OFFSET(MonoLMF, pregs[0]));
+			    MONO_STRUCT_OFFSET(MonoLMF, pregs));
 
-		mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, pregs[0]), SLOT_NOREF);
-		mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, pregs[1]), SLOT_NOREF);
-		mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, pregs[2]), SLOT_NOREF);
-		mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, pregs[3]), SLOT_NOREF);
-		mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, pregs[4]), SLOT_NOREF);
+		for (i = 0; i < 5; i++)
+			mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, pregs) + i * sizeof(gulong), SLOT_NOREF);
 
 		/*---------------------------------------------------------------*/
 		/* On return from this call r2 have the address of the &lmf	 */
@@ -5657,7 +5654,7 @@ mono_arch_emit_prolog (MonoCompile *cfg)
 		/* Set lmf.lmf_addr = jit_tls->lmf				 */	
 		/*---------------------------------------------------------------*/	
 		s390_stg   (code, s390_r2, 0, s390_r13, 				
-			    G_STRUCT_OFFSET(MonoLMF, lmf_addr));			
+			    MONO_STRUCT_OFFSET(MonoLMF, lmf_addr));			
 		mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, lmf_addr), SLOT_NOREF);
 											
 		/*---------------------------------------------------------------*/	
@@ -5674,7 +5671,7 @@ mono_arch_emit_prolog (MonoCompile *cfg)
 		/* Have our lmf.previous_lmf point to the last lmf		 */	
 		/*---------------------------------------------------------------*/	
 		s390_stg   (code, s390_r0, 0, s390_r13, 				
-			    G_STRUCT_OFFSET(MonoLMF, previous_lmf));			
+			    MONO_STRUCT_OFFSET(MonoLMF, previous_lmf));			
 		mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, previous_lmf), SLOT_NOREF);
 											
 		/*---------------------------------------------------------------*/	
@@ -5682,15 +5679,15 @@ mono_arch_emit_prolog (MonoCompile *cfg)
 		/*---------------------------------------------------------------*/	
 		S390_SET   (code, s390_r1, method);
 		s390_stg   (code, s390_r1, 0, s390_r13, 				
-			    G_STRUCT_OFFSET(MonoLMF, method));				
+			    MONO_STRUCT_OFFSET(MonoLMF, method));				
 		mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, method), SLOT_NOREF);
 										
 		/*---------------------------------------------------------------*/	
 		/* save the current IP						 */	
 		/*---------------------------------------------------------------*/	
-		s390_stg   (code, STK_BASE, 0, s390_r13, G_STRUCT_OFFSET(MonoLMF, ebp));
+		s390_stg   (code, STK_BASE, 0, s390_r13, MONO_STRUCT_OFFSET(MonoLMF, ebp));
 		s390_basr  (code, s390_r1, 0);
-		s390_stg   (code, s390_r1, 0, s390_r13, G_STRUCT_OFFSET(MonoLMF, eip));	
+		s390_stg   (code, s390_r1, 0, s390_r13, MONO_STRUCT_OFFSET(MonoLMF, eip));	
 		mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, ebp), SLOT_NOREF);
 		mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, eip), SLOT_NOREF);
 											
@@ -5698,23 +5695,14 @@ mono_arch_emit_prolog (MonoCompile *cfg)
 		/* Save general and floating point registers			 */	
 		/*---------------------------------------------------------------*/	
 		s390_stmg  (code, s390_r2, s390_r12, s390_r13, 				
-			    G_STRUCT_OFFSET(MonoLMF, gregs[2]));			
-		mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, gregs[0]), SLOT_NOREF);
-		mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, gregs[1]), SLOT_NOREF);
-		mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, gregs[2]), SLOT_NOREF);
-		mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, gregs[3]), SLOT_NOREF);
-		mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, gregs[4]), SLOT_NOREF);
-		mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, gregs[5]), SLOT_NOREF);
-		mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, gregs[6]), SLOT_NOREF);
-		mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, gregs[7]), SLOT_NOREF);
-		mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, gregs[8]), SLOT_NOREF);
-		mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, gregs[9]), SLOT_NOREF);
-		mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, gregs[10]), SLOT_NOREF);
+			    MONO_STRUCT_OFFSET(MonoLMF, gregs) + 2 * sizeof(gulong));	
+		for (i = 0; i < 11; i++)
+			mini_gc_set_slot_type_from_fp (cfg, lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, gregs) + i * sizeof(gulong), SLOT_NOREF);
 
-		fpOffset = lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, fregs[0]);
+		fpOffset = lmfOffset + MONO_STRUCT_OFFSET (MonoLMF, fregs);
 		for (i = 0; i < 16; i++) {						
 			s390_std  (code, i, 0, s390_r13, 				
-				   G_STRUCT_OFFSET(MonoLMF, fregs[i]));			
+				   MONO_STRUCT_OFFSET(MonoLMF, fregs) + i * sizeof(gulong));
 			mini_gc_set_slot_type_from_fp (cfg, fpOffset, SLOT_NOREF);
 			fpOffset += sizeof(double);
 		}									
@@ -5723,7 +5711,7 @@ mono_arch_emit_prolog (MonoCompile *cfg)
 		/* Restore the parameter registers now that we've set up the lmf */
 		/*---------------------------------------------------------------*/
 		s390_lmg   (code, s390_r2, s390_r6, s390_r13, 				
-			    G_STRUCT_OFFSET(MonoLMF, pregs[0]));			
+			    MONO_STRUCT_OFFSET(MonoLMF, pregs));	
 	}
 
 	if (cfg->method->save_lmf)

